### PR TITLE
Addition of TanhExp activation function - issue #45929

### PR DESCRIPTION
**Pull request by sangireddysiva**
Addition of TanhExp activation function based on issue #45929 that leads changes to tensorflow/python/keras/activations.py and tensorflow/python/ops/nn_impl.py

### CODE REVIEW

**Comment by sangireddysiva** [comment]
We are not sure what exactly oneDNN unit tests failed checks are talking about. I could see failure in compiling programs and some functionality tests. are these failed checks meant for our code changes?

**Comment by ematejska** [comment]
The API owners are moving to a rotation based system instead of the meetings. Looks like @fchollet is already a reviewer and adding @manivaradarajan. Please review on behalf of the API Owners.

File: tensorflow/python/keras/activations.py
Comment (manivaradarajan) [inline tensorflow/python/keras/activations.py:375]: Rephrase to "TanhExp activation function" for consistency. Please fix throughout.

**Review by manivaradarajan — CHANGES_REQUESTED** [review]
Overall ok. Addressing API documentation only. Please wait for fchollet's review.

**Review by sangireddysiva — COMMENTED** [review]
Rephrased "TanhExponential" to "TanhExp" for consistency

File: tensorflow/python/keras/activations.py
Comment (sangireddysiva) [reply tensorflow/python/keras/activations.py]: updated activations.py with suggested changes

**Review by manivaradarajan — APPROVED** [review]

**Comment by gbaned** [comment]
@sangireddysiva  Can you please address Ubuntu Sanity errors? Thanks!

**Comment by sangireddysiva** [comment]
> @sangireddysiva Can you please address Ubuntu Sanity errors? Thanks!

@gbaned , Adjusted the documentation to address the Ubuntu Sanity errors

**Comment by sangireddysiva** [comment]
@gbaned  , Added Pylint disable line too long statement to address the Ubuntu Sanity errors

File: tensorflow/python/keras/activations.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/activations.py:375]: Would be better to reformat the comment to not have lines that are too long.

**Comment by sangireddysiva** [comment]
@gbaned , fixed Ubuntu CPU Doctest errors, thanks

File: tensorflow/python/keras/activations.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/activations.py:380]: Space above and below this line

File: tensorflow/python/ops/nn_impl.py
Comment (mihaimaruseac) [inline tensorflow/python/ops/nn_impl.py:557]: Reformat comment instead of this

File: tensorflow/python/ops/nn_impl.py
Comment (mihaimaruseac) [inline tensorflow/python/ops/nn_impl.py:562]: Space above this line

File: tensorflow/python/ops/nn_impl.py
Comment (mihaimaruseac) [inline tensorflow/python/ops/nn_impl.py:565]: Space above this line

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by sangireddysiva** [comment]
@mihaimaruseac addresses the requested changes to the documentation, thanks

**Review by mihaimaruseac — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by sangireddysiva** [comment]
@gbaned , There are still some failed checks, do we need to look into them before merging? Thanks,

**Comment by gbaned** [comment]
@mihaimaruseac Can you please assist on above comments from @sangireddysiva. Thanks!

**Comment by rthadur** [comment]
approved by @tensorflow/api-owners internally , trying again to merge this internally

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
You need to regenerate the API goldens as you changed the API. That's why Ubuntu CPU fails and that's why this does not land. See https://source.cloud.google.com/results/invocations/6ac56c56-69eb-4a88-9020-62ef5fb43a17/log

```
Running tests under Python 3.6.10: /usr/bin/python3
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibility
/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/usr/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

E0512 17:59:28.341844 140552551331584 api_compatibility_test.py:293] TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

ERROR:tensorflow:2 differences found between API and golden.
E0512 17:59:28.342252 140552551331584 api_compatibility_test.py:294] 2 differences found between API and golden.
ERROR:tensorflow:    Change detected in python object: tensorflow.keras.activations.
E0512 17:59:28.342407 140552551331584 api_compatibility_test.py:315]     Change detected in python object: tensorflow.keras.activations.
ERROR:tensorflow:
  path: "tensorflow.keras.activations"
  tf_module {
    member_method {
....
```

**Comment by gbaned** [comment]
@sangireddysiva Can you please check @mihaimaruseac's comments and keep us posted ? Thanks!

**Comment by haifeng-jin** [comment]
I am closing the PR.
We are trying minimize the redundancy between the core tf API and the keras API.
You can always import it from the tf APIs.

**Comment by haifeng-jin** [comment]
@sangireddysiva To minimize the redundancy between the Keras API and the tensorflow ops,
the best solution should be remove the Keras layer but keep the tf ops.

**Comment by fchollet** [comment]
@gbaned this PR adds a new activation function in `tf.nn`. Let's ask a TF core team member to review it.

**Comment by gbaned** [comment]
> @gbaned this PR adds a new activation function in `tf.nn`. Let's ask a TF core team member to review it.

@fchollet Thank you for the update.

File: tensorflow/python/ops/nn_impl.py
Comment (allenlavoie) [inline tensorflow/python/ops/nn_impl.py:584]: Why not `tf.recompute_grad`? That will save a bit more memory executing eagerly and avoids repeating the gradient definitions.

Please also add doctested examples to the tf.nn symbol and a unit test. If you're defining a custom gradient it'd need numeric gradient checking too (probably fine to omit for recompute_grad).

**Comment by gbaned** [comment]
@sangireddysiva  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by qlzh727** [comment]
Removing myself as reviewer since the Keras code has been moved to github.com/keras-team/keras. Please split the keras change into a new PR and sent it to keras-team, if you are still interested in contributing this change. Thanks.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**
